### PR TITLE
Improve editing finalized forms [Part 3]

### LINKED ISSUE #6715: It is possible to edit forms that use entities during form update
#### ODK Collect version
master version cb92cab2a2d5a1150e7c29b10676b12fa425a101
#### Android version
10, 15
#### Device used
Redmi 7, Pixel 6a
#### Problem description
It is possible to edit forms that use entities during form update
#### Steps to reproduce the problem
1. Scan user "edity" from https://test.getodk.cloud/projects/513/app-users
2. Go to trees registration editable form and send it 
3. On central in form access add some other forms to user "edity"
4. Go to start new form and refresh list of forms
5. Quickly go to sent, open previously sent form and press pencil icon
#### Expected behavior
Editing forms that use entities should be blocked during form update

### LINKED ISSUE #6716: Resetting saved forms when there's an edit to a form crashes the app
#### ODK Collect version
the master version cb92cab2a2d5a1150e7c29b10676b12fa425a101
#### Android version
10, 14
#### Device used
Redmi 9T, Pixel 7a
#### Problem description
Resetting saved forms when there's an edit to a form crashes the app
The issue doesn't occur in the store version 2025.1.4.

Image

#### Steps to reproduce the problem
1. Download a project with an editable form. (e.g. user test1 in https://test.getodk.cloud/projects/616/app-users)
2. Finalize or send an editable form.
3. Make an edit to the form and finalize it or save it as draft.
4. Go to Settings- Project Management- Reset, select "Saved forms..." and tap "Reset.
#### Expected behavior
I'm not sure what's the expected - should it be possible to reset saved forms when there are edits?

### LINKED ISSUE #6719: Sending finalized editable forms in different order
#### ODK Collect version
the master version cfe07da267470a0c8c8871074c01b19ee6da0127
#### Android version
10, 14
#### Device used
Redmi 9T, Pixel 7a
#### Problem description
If there's a finalized editable form with edits and a user tries to send finalized forms in different order than sending of following edits fails.

https://github.com/user-attachments/assets/72ade2bf-bbbe-4985-a37b-658835c8f66c

#### Steps to reproduce the problem
1. Download a project with an editable form (e.g. https://test.getodk.cloud/projects/616)
2. Set auto send to "off".
3. Finalize an editable form.
4. Create 2 finalized edits.
5. In settings change the URL so that it's wrong (e.g. add a letter).
6. Go to "Ready to send" and select "original" form or the first edit and try to send it (subtext changes- time changes).
7. Tap "Select all" and try to send all.
#### Expected behavior
Instances should be sent according to their finalization time.

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #6715
Closes #6716
Closes #6719

#### Why is this the best possible solution? Were any other approaches considered?
#6715
This issue wasn’t introduced by recent changes (edits), but it has become more noticeable so let’s address it. The problem is that the form lock was always released when exiting a form, regardless of whether the form had actually acquired it. Only forms with entities acquire the lock so only those forms should release it. The fact that we did that always led to the following scenario:

1. Start downloading forms (this acquires the lock) by refreshing the list of blank forms.
2. Open a form that doesn't acquire the lock (i.e., one without entities) and close it quickly.
3. Open a form with entities - this succeeds, even though the form download process is still ongoing and should still hold the lock.

#6716
Here, I've corrected the order in which forms are removed during a full reset to prevent any issues.

#6719 
As proposed in https://github.com/getodk/collect/pull/6709#issuecomment-2894360017, I believe we should send forms based on their finalization time to ensure the order remains consistent, even if sending one of the forms fails. To support this, I added a new column to the instances database.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
The biggest risk relates to #6719. As mentioned earlier, I introduced a new column to track the finalization time. This means upgrading from v2025.1.5 must be carefully tested. During the upgrade, the new column should be added, and for instances that already exist (from v2025.1.5), it should be populated (using `lastStatusChangedDate`) if their status indicates they are finalized - that is, `complete`, `submitted`, or `submission_failed`.

#### Do we need any specific form for testing your changes? If so, please attach one.
Any form that is editable (`client-editable="true"`).

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] added any new strings with date formatting to `DateFormatsTest`
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryViewModel.java [hunk 1]
```diff
@@ -368,7 +368,9 @@ private void updateIndex(boolean isAsync, @Nullable FailedValidationResult valid
     public void exit() {
         formSessionRepository.clear(sessionId);
         ReferenceManager.instance().reset();
-        changeLocks.getFormsLock().unlock();
+        if (form != null && form.usesEntities()) {
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryViewModel.java:371]: Is there a possible state where `form == null` and we still need to unlock?

File: collect_app/src/test/java/org/odk/collect/android/formentry/FormEntryViewModelTest.java [hunk 2]
```diff
@@ -449,4 +451,17 @@ public void isFormEditableAfterFinalization_returnsTrue_whenClientEditableAttrib
 
         assertThat(viewModel.isFormEditableAfterFinalization(), equalTo(true));
     }
+
+    @Test
+    public void exit_releasesFormsLockOnlyIfFormHasEntities() {
+        changeLocks.getFormsLock().lock();
+
+        when(form.usesEntities()).thenReturn(false);
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/formentry/FormEntryViewModelTest.java]: Couldn't you just set a new form session with a real `Form` rather than using Mockito here?

File: forms-test/src/main/java/org/odk/collect/formstest/InMemInstancesRepository.java [hunk 3]
```diff
@@ -120,6 +120,13 @@ public List<Instance> getAllNotDeletedByFormIdAndVersion(String formId, String v
     @Override
     public void delete(Long id) {
         Instance instance = get(id);
+
+        boolean hasEdits = instances.stream()
```
Comment (seadowg) [inline forms-test/src/main/java/org/odk/collect/formstest/InMemInstancesRepository.java:124]: Are we missing a test in `InstancesRepostitoryTest` that would have required this?

File: collect_app/src/main/java/org/odk/collect/android/database/DatabaseObjectMapper.kt [hunk 4]
```diff
@@ -162,6 +165,15 @@ object DatabaseObjectMapper {
             .formVersion(cursor.getString(jrVersionColumnIndex))
             .status(cursor.getString(statusColumnIndex))
             .lastStatusChangeDate(cursor.getLong(lastStatusChangeDateColumnIndex))
+            .finalizationDate(
+                if (cursor.isNull(finalizationDateColumnIndex)) {
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/database/DatabaseObjectMapper.kt]: We should add a `getLongOrNull` to `CursorExt` to handle this logic.

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/instancemanagement/AutoSendTest.kt [hunk 5]
```diff
@@ -211,7 +212,9 @@ class AutoSendTest {
     }
 
     @Test
-    fun whenFormHasAutoSend_formsAreSentInOldestFirstOrder() {
+    fun whenFormHasAutoSend_formsAreSentInOldestFirstOrderBasedOnFinalizationTime() {
```
Comment (seadowg) [inline collect_app/src/androidTest/java/org/odk/collect/android/feature/instancemanagement/AutoSendTest.kt:215]: It's probably worth adding a comment to the setup in this and the manual sending test to explain why we call `alwaysReturnError`.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/pages/SendFinalizedFormPage.java [hunk 6]
```diff
@@ -55,9 +55,9 @@ public SendFinalizedFormPage selectForm(int index) {
         return this;
     }
 
-    public SendFinalizedFormPage sortByDateNewestFirst() {
```
Comment (seadowg) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/pages/SendFinalizedFormPage.java:58]: Don't think we need to delete this method. We may well need it for another test in the future.

File: collect_app/src/main/java/org/odk/collect/android/database/instances/InstanceDatabaseMigrator.java [hunk 7]
```diff
@@ -151,6 +152,11 @@ private void upgradeToVersion8(SQLiteDatabase db) {
     private void upgradeToVersion9(SQLiteDatabase db) {
         db.execSQL("ALTER TABLE " + INSTANCES_TABLE_NAME + " ADD COLUMN " + EDIT_OF + " integer REFERENCES " + INSTANCES_TABLE_NAME + "(" + _ID + ") CHECK (" + EDIT_OF + " != " + _ID + ")");
         db.execSQL("ALTER TABLE " + INSTANCES_TABLE_NAME + " ADD COLUMN " + EDIT_NUMBER + " integer CHECK ((" + EDIT_OF + " IS NULL AND " + EDIT_NUMBER + " IS NULL) OR + (" + EDIT_OF + " IS NOT NULL AND + " + EDIT_NUMBER + " IS NOT NULL))");
+        db.execSQL("ALTER TABLE " + INSTANCES_TABLE_NAME + " ADD COLUMN " + FINALIZATION_DATE + " date");
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/database/instances/InstanceDatabaseMigrator.java:160]: Is version 9 not already out in a beta? I might be wrong. If not though, this will need to be a version 10 migration.

**Review by seadowg — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryViewModel.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryViewModel.java:371]: Theoretically, it might be possible... something like this:

1. A form that uses entities is opened, acquiring a form lock.
2. The form is somehow deleted while it’s still open.
3. When the user exits, the lock isn’t released.

However, I’m not sure whether this would result in a crash at some point in the process.
It seems like such an edge case that I don’t think we need to handle it.

File: collect_app/src/test/java/org/odk/collect/android/formentry/FormEntryViewModelTest.java (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/test/java/org/odk/collect/android/formentry/FormEntryViewModelTest.java]: Yes I can do that.

File: collect_app/src/main/java/org/odk/collect/android/database/instances/InstanceDatabaseMigrator.java (hunk 7 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/database/instances/InstanceDatabaseMigrator.java:160]: Right, done.

**Review by seadowg — APPROVED** [review]

**Comment by WKobus** [comment]
**Tested with success**

Verified on device with Android 15

**Verified cases:**
- Deleting and resetting saved forms
- Opening edits of entity form while form update is in progress
- Sending finalized edits in different order
- Upgrading app
- Regression check on editing finalized forms

**Comment by dbemke** [comment]
**Tested with success**

Verified on device with Android 10
